src/ucp/core/ucp_worker.c

```diff
@@ -533,65 +454,72 @@ ucs_status_t ucp_worker_set_ep_failed(ucp_worker_h worker, ucp_ep_h ucp_ep,
-                                      err_handle_arg, UCS_CALLBACKQ_FLAG_ONESHOT,
-                                      &prog_id);
-
-    if ((ucp_ep_ext_control(ucp_ep)->err_cb == NULL) &&
-        (ucp_ep->flags & UCP_EP_FLAG_USED)) {
-        /* do not print error if connection reset by remote peer since it can
-         * be part of user level close protocol  */
-        log_level = (status == UCS_ERR_CONNECTION_RESET) ? UCS_LOG_LEVEL_DIAG :
-                    UCS_LOG_LEVEL_ERROR;
+    /* The EP can be closed from last completion callback */
+    ucp_ep_discard_lanes(ucp_ep, status);
+    ucp_ep_reqs_purge(ucp_ep, status);
+    ucp_stream_ep_cleanup(ucp_ep);
 
-        if (lane != UCP_NULL_LANE) {
-            if (lane == ucp_ep_get_cm_lane(ucp_ep)) {
-                ucs_strncpy_safe(lane_info_str, "CM lane",
-                                 sizeof(lane_info_str));
+    if (ucp_ep->flags & UCP_EP_FLAG_USED) {
+        if (ucp_ep->flags & UCP_EP_FLAG_CLOSE_REQ_VALID) {
+            ucs_assert(ucp_ep->flags & UCP_EP_FLAG_CLOSED);
+            /* Promote close operation to CANCEL in case of transport error,
+             * since the disconnect event may never arrive. */
+            close_req                        = ucp_ep_ext_control(ucp_ep)->
+                close_req.req;
+            close_req->send.flush.uct_flags |= UCT_FLUSH_FLAG_CANCEL;
+            ucp_ep_local_disconnect_progress(close_req);
+        } else if (ucp_ep_ext_control(ucp_ep)->err_cb == NULL) {
+            /* Do not print error if connection reset by remote peer since it
+             * can be part of user level close protocol */
+            log_level = (status == UCS_ERR_CONNECTION_RESET) ?
+                        UCS_LOG_LEVEL_DIAG : UCS_LOG_LEVEL_ERROR;
+
+            if (lane != UCP_NULL_LANE) {
+                if (lane == ucp_ep_get_cm_lane(ucp_ep)) {
+                    ucs_strncpy_safe(lane_info_str, "CM lane",
+                                     sizeof(lane_info_str));
+                } else {
+                    rsc_index = ucp_ep_get_rsc_index(ucp_ep, lane);
+                    tl_rsc    = &worker->context->tl_rscs[rsc_index].tl_rsc;
+
+                    ucs_snprintf_safe(lane_info_str, sizeof(lane_info_str),
+                                      UCT_TL_RESOURCE_DESC_FMT,
+                                      UCT_TL_RESOURCE_DESC_ARG(tl_rsc));
+                }
             } else {
-                rsc_index = ucp_ep_get_rsc_index(ucp_ep, lane);
-                tl_rsc    = &worker->context->tl_rscs[rsc_index].tl_rsc;
-
-                ucs_snprintf_safe(lane_info_str, sizeof(lane_info_str),
-                                  UCT_TL_RESOURCE_DESC_FMT,
-                                  UCT_TL_RESOURCE_DESC_ARG(tl_rsc));
+                ucs_assert(uct_ep == NULL);
+                ucs_strncpy_safe(lane_info_str, "wireup lane",
+                                 sizeof(lane_info_str));
             }
```

Comment:
maybe move to separate func, e.g. `ucp_ep_get_lane_info_str`?

Reply:
good idea, done